packages/gatsby-remark-images/src/index.js

```diff
@@ -412,62 +416,164 @@ module.exports = (
+                        }
+                      }
+
+                      return resolve()
+                    })
                 )
+              )
+
+              if (!valueChanged) {
+                // No need to dump AST
+                return resolve()
+              }
+
+              let rawHTML = generate(ast, {}, node.value).code
+
+              // Strip the final semicolon
+              rawHTML = rawHTML.slice(0, rawHTML.length - 1)
+
+              // Replace the image node with an inline HTML node.
+              if (node.type !== `jsx`) {
+                node.type = `html`
+              }
+              node.value = rawHTML
+
+              return resolve(node)
+            }
 
-                if (rawHTML) {
-                  // Replace the image string
-                  thisImg.replaceWith(rawHTML)
-                } else {
+            const useCheerio = async () => {
+              const $ = cheerio.load(node.value)
+              if ($(`img`).length === 0) {
+                // No img tags
+                return resolve()
+              }
+
+              let imageRefs = []
+              $(`img`).each(function() {
+                imageRefs.push($(this))
+              })
+
+              for (let thisImg of imageRefs) {
+                // Get the details we need.
+                let formattedImgTag = {}
+                formattedImgTag.url = thisImg.attr(`src`)
+                formattedImgTag.title = thisImg.attr(`title`)
+                formattedImgTag.alt = thisImg.attr(`alt`)
+
+                if (!formattedImgTag.url) {
                   return resolve()
                 }
+
+                const fileType = getImageInfo(formattedImgTag.url).ext
+
+                // Ignore gifs as we can't process them,
+                // svgs as they are already responsive by definition
+                if (
+                  isRelativeUrl(formattedImgTag.url) &&
+                  fileType !== `gif` &&
+                  fileType !== `svg`
+                ) {
+                  const rawHTML = await generateImagesAndUpdateNode(
+                    formattedImgTag,
+                    resolve,
+                    inLink
+                  )
+
+                  if (rawHTML) {
+                    // Replace the image string
+                    thisImg.replaceWith(rawHTML)
+                  } else {
+                    return resolve()
+                  }
+                }
               }
-            }
 
-            // Replace the image node with an inline HTML node.
-            node.type = `html`
-            node.value = $(`body`).html() // fix for cheerio v1
+              // Replace the image node with an inline HTML node.
+              node.type = `html`
+              node.value = $(`body`).html() // fix for cheerio v1
+
+              return resolve(node)
+            }
 
-            return resolve(node)
+            return useBabel().catch(() => useCheerio())
```

Comment:
given the potential perf hit (we should benchmark) - would it be feasible to just check `node.type` and if it's `html` use cheerio and if it's `jsx` use babel?

Reply:
We should use something like https://www.npmjs.com/package/benchmark to check this vs current implementation using input that works now with cheerio (so using some remark AST fixture as input - likely couple of those). I'm very not comfortable with this change as-is

Reply:
Oh right, I haven't even thought about that 👍

Reply:
Done

Reply:
I don't know if MDX always sets HTML blocks as JSX, or if it somehow distinguishes.

Reply:
Ah I don't know either. @johno halp ;)

Reply:
as for "malformatted" jsx part - wouldn't gatsby-plugin-mdx throw exception even before calling `gatsby-remark-images` if there was malfortmatted input that mdx can't handle?

Reply:
I guess MDX would first try to parse any block as JSX, and if it fails it considers the block as regular text. My supposition is based on this (supposedly working) snippet:

```md
---
title: Hello World
---

import FooBar from 'foobar';

Hello world!

<FooBar />
```

Then it would make sense to consider that blocks of type `jsx` would hold valid JSX, and therefore to only parse them using Babel.

Reply:
Bump @johno